sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala

```diff
@@ -137,28 +137,39 @@ class SQLQueryTestSuite extends QueryTest with SharedSQLContext {
     }
   }
 
+  // For better test coverage, runs the tests on mixed config sets: WHOLESTAGE_CODEGEN_ENABLED
+  // and CODEGEN_FACTORY_MODE.
+  private lazy val codegenConfigSets = Array(
+    ("true", "CODEGEN_ONLY"),
+    ("false", "CODEGEN_ONLY"),
+    ("false", "NO_CODEGEN")
+  ).map { case (wholeStageCodegenEnabled, codegenFactoryMode) =>
+    Array(SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> wholeStageCodegenEnabled,
+      SQLConf.CODEGEN_FACTORY_MODE.key -> codegenFactoryMode)
+  }
+
   /** Run a test case. */
   private def runTest(testCase: TestCase): Unit = {
     val input = fileToString(new File(testCase.inputFile))
 
     val (comments, code) = input.split("\n").partition(_.startsWith("--"))
 
-    // Runs all the tests on both codegen-only and interpreter modes
-    val codegenConfigSets = Array(CODEGEN_ONLY, NO_CODEGEN).map {
-      case codegenFactoryMode =>
-        Array(SQLConf.CODEGEN_FACTORY_MODE.key -> codegenFactoryMode.toString)
-    }
-    val configSets = {
-      val configLines = comments.filter(_.startsWith("--SET")).map(_.substring(5))
-      val configs = configLines.map(_.split(",").map { confAndValue =>
-        val (conf, value) = confAndValue.span(_ != '=')
-        conf.trim -> value.substring(1).trim
-      })
-      // When we are regenerating the golden files, we don't need to set any config as they
-      // all need to return the same result
-      if (regenerateGoldenFiles) {
-        Array.empty[Array[(String, String)]]
-      } else {
+    // List of SQL queries to run
+    // note: this is not a robust way to split queries using semicolon, but works for now.
+    val queries = code.mkString("\n").split("(?<=[^\\\\]);").map(_.trim).filter(_ != "").toSeq
+
+    // When we are regenerating the golden files, we don't need to set any config as they
```

Comment:
I am not sure about this. Imagine a case which is failing without setting a config for instance: I think we should pick one config set (any is fine) instead.

Reply:
> Imagine a case which is failing without setting a config for instance

We will check the result for all different configs, and test will fail, right?

Reply:
> We will check the result for all different configs

For all different configs specified there doesn't mean without setting that config at all. For instance, let's imagine that we add a config for throwing an exception when a decimal overflows (with this behavior being the default): in the current tests, we can just add the config value to "do not throw an exception" for all the config sets which are present there. After this patch, that wouldn't work.

Reply:
Currently this test framework requires all the cases return the same result(without config and with any combination of configs). Did I miss something?

Reply:
I think only with any combination of configs is true, not without config.

Reply:
If we do need to test the SQL strict mode, I think we need to adopt SPARK-24562 then.

Reply:
I mean in general, in the current condition, there is no way you can successfully run the tests if the default value of a config produces (for any reason) an output different from some other values and we want to test only the non-default values. It may not be a problem but I think anyway this is a limitation which we can easily avoid by adding here the `SET`s for the first config set if any. So I see no reason why not to do that, which is safer IMHO.

Reply:
I think it's easier to reason about the test if we require it to always produce the same result without config and with any combination of configs.

If no tests depend on it, I'd like to not bother about it and keep it as it was. We already clear configs with generating result: https://github.com/apache/spark/pull/23213/files#diff-432455394ca50800d5de508861984ca5L157